### PR TITLE
Add python AWS Signature v4 support

### PR DESCRIPTION
**Pull request by jblakeney**
Work to implement feature in issue #56 

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-json-schema-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-json-schema-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/python*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).

### CODE REVIEW

File: modules/openapi-json-schema-generator/src/test/resources/3_0/python/petstore-with-fake-endpoints-models-for-testing-with-aws-sigv4.yaml [hunk 1]
```diff
@@ -0,0 +1,265 @@
+openapi: 3.0.0
```
Comment (spacether) [inline modules/openapi-json-schema-generator/src/test/resources/3_0/python/petstore-with-fake-endpoints-models-for-testing-with-aws-sigv4.yaml:1]: If you are making your own spec for an example can you cut it down to one endpoint with the new security scheme and one without it?

Comment (jblakeney) [reply modules/openapi-json-schema-generator/src/test/resources/3_0/python/petstore-with-fake-endpoints-models-for-testing-with-aws-sigv4.yaml:1]: Good idea, I had just copied a chunk of the original one, but had not paired it down to what I needed.

I have now added a test for the `get` call without the security set, however I am getting a strange issue. I was wondering if you could shed some light on it. 

I copied the test from https://github.com/openapi-json-schema-tools/openapi-json-schema-generator/blob/master/samples/openapi3/client/petstore/python/tests_manual/test_paths/test_pet_pet_id/test_get.py

However for some reason the model isn't being generated with the paired down spec, however in both paths it still references the Pet schema as a response type. Instead I am getting a `DynamicSchema` object type as a response.

Comment (spacether) [reply modules/openapi-json-schema-generator/src/test/resources/3_0/python/petstore-with-fake-endpoints-models-for-testing-with-aws-sigv4.yaml:1]: What do you mean by being generated? Files being written? DynamicSchema instances are expected. Validated instances are made of dynamic classes (DynamicSchema) that subclass the validated classes.

**Comment by spacether** [comment]
Why are CI tests not running on this PR?

**Comment by spacether** [comment]
Closing and re opening to kick off ci

File: modules/openapi-json-schema-generator/src/test/resources/3_0/python/petstore-with-fake-endpoints-models-for-testing-with-aws-sigv4.yaml (hunk 1 above)
Comment (jblakeney) [reply modules/openapi-json-schema-generator/src/test/resources/3_0/python/petstore-with-fake-endpoints-models-for-testing-with-aws-sigv4.yaml:1]: Yes so when I generated the sample, it didnt make the Pet Model. As you can see below trying to reference it in the test fails, however that schema is defined in the openapi yaml.

```
ImportError while importing test module '/home/jblakeney/work/dev/openapi-json-schema-generator/samples/openapi3/client/petstore/python_aws_sigv4/tests_manual/test_paths/test_pet_pet_id/test_get.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests_manual/test_paths/test_pet_pet_id/test_get.py:16: in <module>
    from petstore_api.model.pet import Pet
E   ModuleNotFoundError: No module named 'petstore_api.model'
```

Comment (spacether) [reply modules/openapi-json-schema-generator/src/test/resources/3_0/python/petstore-with-fake-endpoints-models-for-testing-with-aws-sigv4.yaml:1]: You need to rebase on the 2_0_0 branch. Models are not there. There are in components.schema

Comment (jblakeney) [reply modules/openapi-json-schema-generator/src/test/resources/3_0/python/petstore-with-fake-endpoints-models-for-testing-with-aws-sigv4.yaml:1]: Ahh perfect, that fixes it. I should of looked at the test in the code, and not just the link you provided as that was the PR! My bad!

File: bin/configs/python_aws_sigv4.yaml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+generatorName: python
+outputDir: samples/openapi3/client/petstore/python_aws_sigv4
```
Comment (spacether) [inline bin/configs/python_aws_sigv4.yaml:2]: How about renaming the spec file something like signing_spec.yaml?
How about outputting the sample into the features directory?
How about changing the package name to something like signing_spec_api?
This sample is not related to petstore.

File: modules/openapi-json-schema-generator/src/main/java/org/openapitools/codegen/languages/PythonClientCodegen.java [hunk 3]
```diff
@@ -455,7 +455,10 @@ public void processOpts() {
                 (openAPI.getComponents() != null ? openAPI.getComponents().getSecuritySchemes() : null) : null;
         List<CodegenSecurity> authMethods = fromSecurity(securitySchemeMap);
         if (ProcessUtils.hasHttpSignatureMethods(authMethods)) {
-            supportingFiles.add(new SupportingFile("signing." + templateExtension, packagePath(), "signing.py"));
+            supportingFiles.add(new SupportingFile("signing_http_signing." + templateExtension, packagePath() + File.separatorChar + "signing", "http_signing.py"));
+        }
+        if (ProcessUtils.hasAwsSignatureV4Methods(authMethods)) {
+            supportingFiles.add(new SupportingFile("signging_aws_sigv4." + templateExtension, packagePath() + File.separatorChar + "signing", "aws_sigv4.py"));
```
Comment (spacether) [inline modules/openapi-json-schema-generator/src/main/java/org/openapitools/codegen/languages/PythonClientCodegen.java:461]: How about also adding an init file to the signing folder?

File: modules/openapi-json-schema-generator/src/main/resources/python/rest.handlebars [hunk 4]
```diff
@@ -125,6 +128,11 @@ class RESTClientObject(object):
                   len(timeout) == 2):
                 timeout = urllib3.Timeout(connect=timeout[0], read=timeout[1])
 
+        {{#if hasAwsSignatureV4Methods}}
+        if self.aws_sigv4_info:
+            headers.extend(self.aws_sigv4_info.get_aws_signature_v4_headers(method, url, body))
```
Comment (spacether) [inline modules/openapi-json-schema-generator/src/main/resources/python/rest.handlebars:133]: This code should be route specific.
How about moving this code into update_params_for_auth
And moving the invocation of that call after the header extension?
https://github.com/openapi-json-schema-tools/openapi-json-schema-generator/blob/master/modules/openapi-json-schema-generator/src/main/resources/python/api_client.handlebars#L1063

Comment (jblakeney) [reply modules/openapi-json-schema-generator/src/main/resources/python/rest.handlebars:133]: Sorry was taking a break all weekend so just getting back to this! 

I like the idea of it being per route, however there is a comment above where used_headers.update is calling saying it should be called after any auth settings. Does that not apply anymore?

File: modules/openapi-json-schema-generator/src/main/java/org/openapitools/codegen/languages/PythonClientCodegen.java (hunk 3 above)
Comment (jblakeney) [reply modules/openapi-json-schema-generator/src/main/java/org/openapitools/codegen/languages/PythonClientCodegen.java:461]: I had added the template, but forget to add it as a supporting file! Good catch!

File: modules/openapi-json-schema-generator/src/main/resources/python/rest.handlebars (hunk 4 above)
Comment (spacether) [reply modules/openapi-json-schema-generator/src/main/resources/python/rest.handlebars:133]: No worries. Also I am on vacation so forgive some delayed responses from me here. It kind of doesn't apply. I wrote it. The default headers must come first. Update has to happen after, and those are spec defined headers. For your signing and the other security scheme they must come last because they depend upon the payload in general, right?
For the other security schemes, the Autherization header is disallowed as an explicit header input. So it is probably fine to update all security info last.

**Comment by spacether** [comment]
@jblakeney friendly ping. Do you still want this? I haven't seen any updates in a week.

**Comment by jblakeney** [comment]
> @jblakeney friendly ping. Do you still want this? I haven't seen any updates in a week.

Hey, yes sorry I do plan to finish this. Its been a crazy work week. Ill aim to get the suggestions fixed up this week! Sorry for the delay!

**Comment by spacether** [comment]
No worries. I was just curious. Rebases on the 2_0_0 branch can be painful. If it is too much of a hassle, I find the easiest way to do it is to:
- squash the feature commits in your branch into 1 commit, copy it into a new branch
- locally delete your original branch
- make a new version of your original branch from the latest 2_0_0
- get cherry-pick your commit
- rebuild and regen samples
- force push your branch

**Comment by jblakeney** [comment]
> No worries. I was just curious. Rebases on the 2_0_0 branch can be painful. If it is too much of a hassle, I find the easiest way to do it is to:
> 
> * squash the feature commits in your branch into 1 commit, copy it into a new branch
> * locally delete your original branch
> * make a new version of your original branch from the latest 2_0_0
> * get cherry-pick your commit
> * rebuild and regen samples
> * force push your branch

Awesome ill give that a try, im usually pretty good at doing merge conflicts but we shall see! Ill give that an attempt tomorrow and ill let you know if I run into any issues.

**Comment by jblakeney** [comment]
Sorry for the delay, just wanted to update you here. I got the conflcits resolved on a new branch, gonna make my changes and then ill get it merged into this branch and push it. Aiming to hopefully have that done by this weekend.

**Comment by jblakeney** [comment]
Sorry again, im gonna close with for now. We do still need this but due to work deadlines before Christmas I havent had a chance to dedicate time to fix this up and properly test it. Ill re-open a PR once im ready to go. Sorry again!

**Comment by spacether** [comment]
No worries. When you have time you can add it later
